[quant][graphmode][fx] Standalone module support {input/output}_quantized_idxs

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49754 [quant][graphmode][fx] Standalone module support {input/output}_quantized_idxs**

Summary:
This PR adds the support for {input/output}_quantized_idxs for standalone module.

if input_quantized_idxs = [] and output_quantized_idxs = [], the standalone module will be expecting float
input and produce float output, and will quantize the input and dequantize output internally

if input_quantized_idxs = [0] and otuput_qiuantized_idxs = [0], the standalone module will be expecting quantized
input and produce quantized output, the input will be quantized in the parent module, and output will be dequantized
in the parent module as well, this is similar to current quantized modules like nn.quantized.Conv2d

For more details, please see the test case

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_standalone_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25684692](https://our.internmc.facebook.com/intern/diff/D25684692)

jerryzh168 committed:
[quant][graphmode][fx] Standalone module support {input/output}_quantized_idxs

Summary:
This PR adds the support for {input/output}_quantized_idxs for standalone module.

if input_quantized_idxs = [] and output_quantized_idxs = [], the standalone module will be expecting float
input and produce float output, and will quantize the input and dequantize output internally

if input_quantized_idxs = [0] and otuput_qiuantized_idxs = [0], the standalone module will be expecting quantized
input and produce quantized output, the input will be quantized in the parent module, and output will be dequantized
in the parent module as well, this is similar to current quantized modules like nn.quantized.Conv2d

For more details, please see the test case

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_standalone_module

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -574,45 +577,91 @@ def forward(self, x):
+                    ns.call_module(nnq.Conv2d): 1,
+                    ns.call_method("dequantize") : 1,
+                }
+            else:
```
When inputs and outputs are quantized, why do we expect to see quantize_per_tensor and dequantize?

jerryzh168 replied:
which one are you referring to? in standalone_convert_count_check we didn't expect to see quantize_per_tensor or dequantize. we only expect to see them in parent (convert_count_check)

raghuramank100 replied:
Got it, thanks

raghuramank100 commented on torch/quantization/fx/quantize.py:
```diff
@@ -130,19 +131,22 @@ def insert_observer_for_special_module(
             torch.quantization.quantize_fx._prepare_standalone_module_fx  # type: ignore
         observed_standalone_module = \
-            prepare(standalone_module, standalone_module_qconfig_dict, standalone_prepare_config_dict)
+            prepare(standalone_module, sm_qconfig_dict, sm_prepare_config_dict)
```
For my understanding: Why do we call prepare for standalone modules? Shouldnt we swap a standalone module with a custom module containing the observers?

jerryzh168 replied:
standalone modules are still quantized with our api, user just need to specify which one is standalone module.
it is swapped with a module containing the observers, and this module is generated by calling our _prepare_standalone_module_fx api

jerryzh168 committed:
Update on "[quant][graphmode][fx] Standalone module support {input/output}_quantized_idxs"


Summary:
This PR adds the support for {input/output}_quantized_idxs for standalone module.

if input_quantized_idxs = [] and output_quantized_idxs = [], the standalone module will be expecting float
input and produce float output, and will quantize the input and dequantize output internally

if input_quantized_idxs = [0] and otuput_qiuantized_idxs = [0], the standalone module will be expecting quantized
input and produce quantized output, the input will be quantized in the parent module, and output will be dequantized
in the parent module as well, this is similar to current quantized modules like nn.quantized.Conv2d

For more details, please see the test case

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_standalone_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25684692](https://our.internmc.facebook.com/intern/diff/D25684692)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Standalone module support {input/output}_quantized_idxs"


Summary:
This PR adds the support for {input/output}_quantized_idxs for standalone module.

if input_quantized_idxs = [] and output_quantized_idxs = [], the standalone module will be expecting float
input and produce float output, and will quantize the input and dequantize output internally

if input_quantized_idxs = [0] and otuput_qiuantized_idxs = [0], the standalone module will be expecting quantized
input and produce quantized output, the input will be quantized in the parent module, and output will be dequantized
in the parent module as well, this is similar to current quantized modules like nn.quantized.Conv2d

For more details, please see the test case

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_standalone_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25684692](https://our.internmc.facebook.com/intern/diff/D25684692)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -610,45 +613,92 @@ def forward(self, x):
+            "output_quantized_idxs": [0],  # quantized output
+        }
+        # is_name, is_float
+        options = itertools.product([True, False], [True, False])
```
optional: 4 test cases instead of 1, to make it easier to fix things in the future (so the person fixing a bug doesn't have to comment out the test cases not relevant to the bug)

jerryzh168 replied:
sure

jerryzh168 replied:
will split to two cases since name and class are pretty similar

vkuzo commented on torch/quantization/fx/observed_module.py:
```diff
@@ -35,6 +35,12 @@ def is_observed_module(module: Any) -> bool:
     return isinstance(module, ObservedGraphModule)
 
 class ObservedStandaloneGraphModule(ObservedGraphModule):
+    def get_preserved_attr_names(self):
```
nit: can we annotate the return type?

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -487,14 +512,14 @@ def load_arg(a):
                 if qconfig is not None:
                     assert obj is not None
-                    insert_observer_for_special_module(
+                    standalone_module_input_idxs = insert_observer_for_special_module(
```
optional readability nit (not related to this PR, and optional): maybe we can call the function `maybe_insert_observer_for_special_module`, or move the gating here?

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -631,32 +671,46 @@ def load_arg(quantized: Optional[Union[List[Any], bool, Tuple[Any, ...]]]
-                if quantized is None:
+                # we'll update the format of `quantized`
+                # to better match arg_or_args
+                updated_quantized = quantized
```
nit: can we annotate the expected type?  The logic is not intuitive, having the type would help with readability.

jerryzh168 replied:
sure

jerryzh168 replied:
updated_quantized will have the same type as quantized

jerryzh168 committed:
Update on "[quant][graphmode][fx] Standalone module support {input/output}_quantized_idxs"


Summary:
This PR adds the support for {input/output}_quantized_idxs for standalone module.

if input_quantized_idxs = [] and output_quantized_idxs = [], the standalone module will be expecting float
input and produce float output, and will quantize the input and dequantize output internally

if input_quantized_idxs = [0] and otuput_qiuantized_idxs = [0], the standalone module will be expecting quantized
input and produce quantized output, the input will be quantized in the parent module, and output will be dequantized
in the parent module as well, this is similar to current quantized modules like nn.quantized.Conv2d

For more details, please see the test case

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_standalone_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25684692](https://our.internmc.facebook.com/intern/diff/D25684692)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Standalone module support {input/output}_quantized_idxs"


Summary:
This PR adds the support for {input/output}_quantized_idxs for standalone module.

if input_quantized_idxs = [] and output_quantized_idxs = [], the standalone module will be expecting float
input and produce float output, and will quantize the input and dequantize output internally

if input_quantized_idxs = [0] and otuput_qiuantized_idxs = [0], the standalone module will be expecting quantized
input and produce quantized output, the input will be quantized in the parent module, and output will be dequantized
in the parent module as well, this is similar to current quantized modules like nn.quantized.Conv2d

For more details, please see the test case

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_standalone_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25684692](https://our.internmc.facebook.com/intern/diff/D25684692)

[ghstack-poisoned]

mruberry commented:
Unlanding. This broke several builds with mypy failures. See the Github CI on this PR for examples.
